WooCommerce/Classes/ViewModels/ProductDetailsViewModel.swift

```diff
@@ -0,0 +1,453 @@
+import Foundation
+import Yosemite
+import UIKit
+import Gridicons
+
+
+// MARK: - Product details view model
+//
+final class ProductDetailsViewModel {
+
+    /// Closures
+    ///
+    var onError: (() -> Void)?
+    var onReload: (() -> Void)?
+
+    /// Yosemite.Product
+    ///
+    var product: Product {
+        didSet {
+            reloadTableViewSectionsAndData()
+        }
+    }
+
+    /// Nav bar title
+    ///
+    var title: String {
+        return product.name
+    }
+
+    /// Product ID
+    ///
+    var productID: Int {
+        return product.productID
+    }
+
+    /// Sections to be rendered
+    ///
+    private(set) var sections = [Section]()
+
+    /// EntityListener: Update / Deletion Notifications.
+    ///
+    private lazy var entityListener: EntityListener<Product> = {
```

Comment:
I've only ever seen the EntityListener live in the VC (OrderDetailsVC & NotificationsDetailsVC) with the read-only entity being set on the view model. I don't have opinions either way yet; I'm curious to hear yours!

Reply:
In this case, I think it makes sense to move it to the view model, since the view model is the one that is aware of any data (i.e. a Product)

Reply:
Moving the entityListener to the view model did make the refactor much easier. I tried twice to move code without moving the entityListener and failed.